feat: :sparkles: implement update product function, DTO, unit and int…

ineBallardin opened the pull request:
…egration tests

Resolve issue: #3

ineBallardin committed:
feat: :sparkles: implement update product function, DTO, unit and integration tests

Resolve issue: #3

arthurazevedofranca commented on src/usecase/product/update/update.product.integration.spec.ts:
```diff
@@ -0,0 +1,48 @@
+
+    it("should update a product", async () => {
+        const productRepository = new ProductRepository();
+        const productUsecase = new UpdateProductUseCase(productRepository);
```
acho importante mudar o nome da variavel, productUseCase ta mto generico. pode ser updateProduct sem usecase que induz melhor quem esta lendo

arthurazevedofranca commented on src/usecase/product/update/update.product.integration.spec.ts:
```diff
@@ -0,0 +1,48 @@
+
+        const output = await productUsecase.execute(input);
+
+        expect(output).toEqual(input);
```
acho que o assert verificando se foi atualizado na base meesmo, nao so confiando no retorno do usecase

arthurazevedofranca commented on src/usecase/product/update/update.product.unit.spec.ts:
```diff
@@ -0,0 +1,32 @@
+
+        const output = await productUpdateUseCase.execute(input);
+
+        expect(output).toEqual(input);
```
falta assert verificando se os metodos mockados foram chamados mesmo ja que eh um teste unitario

arthurazevedofranca commented on src/usecase/product/update/update.product.usecase.ts:
```diff
@@ -0,0 +1,28 @@
+        product.changeName(input.name);
+        product.changePrice(input.price);
+
+        await this.ProductRepository.update(product);
```
esse update ja nao deveria retornar o objeto atualizado?

ineBallardin replied:
Na vdd o `ProductRepository` foi feito pra não retornar nenhum valor por padrão. O Wesley até explica o pq em uma das aulas, mas eu esqueci, vou procurar essa aula de nv pra revisar o pq. Veja como foi feita a [interface](https://github.com/ineBallardin/clean_arch_fc/blob/main/src/domain/product/repository/product-repository.interface.ts).

arthurazevedofranca replied:
ai complica ne ZZzzzzz

ineBallardin replied:
simmm

arthurazevedofranca replied:
Outro trecho que vou dar resolve mas vou cobrar nos próximos PRs

arthurazevedofranca commented on src/usecase/product/update/update.product.usecase.ts:
```diff
@@ -0,0 +1,28 @@
+        input: InputUpdateProductDto
+    ): Promise<OutputUpdateProductDto> {
+        const product = await this.ProductRepository.find(input.id);
+        product.changeName(input.name);
```
me parece mais um patch aqui doq um update de tudo ne, se fosse para tudo iria fazer um update where by id e setando todos os valores que vieram

ineBallardin replied:
@arthurazevedofranca sim, pelo q me lembro das aulas é só pra atualizar os campos q forem passados através do `input` msm, não tendo a necessidade de passar todos os campos sempre q for atualizar algum campo. Se não me engano é algo nessa linha, mas vou procurar a aula pra ter certeza.

arthurazevedofranca replied:
entao, muda para um update especifico, mto mais performatico. Nao?

ineBallardin replied:
@arthurazevedofranca se vc diz deve ser. Mas como essa foi uma implementação do Wesley, não sei se eu mexer agr pode afetar o que ele pretende fazer no projeto daqui p frente.

arthurazevedofranca replied:
Tá valendo! Vou marcar como resolve mas vou cobrar nos próximos PRs

ineBallardin committed:
refactor: :recycle: generic 'productUseCase' with 'updateProduct' for improved readability and understanding

Related to issue: #3

ineBallardin committed:
refactor: :white_check_mark: revised test logic to ensure database persistence of updates

Revised test logic to ensure database persistence of updates. Instead of solely comparing the 'execute' method's return to the input, the code now fetches the updated product from the repository post-use case execution, verifying that the updated fields align with the provided input.

Related issue: #3

ineBallardin committed:
test: :white_check_mark: added verification for the invocation of mocked methods to ensure their proper execution

Related to issue: #3

ineBallardin committed:
refactor: :recycle: improve code readability and comprehension by adjusting variables names

Related to issue: #3
